[poro] step 4 컨트롤러 리팩터링, Annotation 자동 URL 매핑 구현

Gwonwoo-Nam opened the pull request:
## 구현 내용
- Controller를 DispatcherServlet처럼 요청이 들어오는 FrontController와 URL마다 요청을 처리할 하위 컨트롤러로 구조를 분리하였습니다.
- FrontController에서는 <URL, Controller>의 map 구조로 url 매핑 정보를 저장하고 있으며, url에 따라 컨트롤러에게 요청이 위임되도록 설계하였습니다.
- RequestMapping이라는 URL을 가지고 있는 Annotation을 만들었고 FrontController에서 리플렉션으로 controller 패키지에 있는 controller 구현체 annotation 정보를 파싱해 등록합니다.

## 고민한 점
- 원래의 조건문으로 분기하는 구조에서는 URL이 늘어날수록 분기문이 늘어날 수 밖에 없는 구조라 Map으로 매핑시켜서 간결하게 만들고자 했습니다.
- Reflection을 사용하니 발생가능한 예외가 많아져서 잘 처리해야할 것 같습니다.
- 가독성은 좋아졌지만 classLoader를 사용해서 파일의 경로에 의존하는 코드가 생겼다는 점이 마음에 걸립니다.

Gwonwoo-Nam committed:
feat(step-5):Controller 구조 Front Controller와 상세 Controller로 구조 분리

- UrlMapper 삭제, Controller 인터페이스 구현 및 FrontController 생성(기존 DispatcherServlet 역할과 유사)

Gwonwoo-Nam committed:
feat(step-5):Controller 구조 Front Controller와 상세 Controller로 구조 분리

- UrlMapper 삭제, Controller 인터페이스 구현 및 FrontController 생성(기존 DispatcherServlet 역할과 유사)

Gwonwoo-Nam committed:
test(step-5):테스트코드 메서드 이름 변경

Gwonwoo-Nam committed:
feat(step-5):Reflection을 이용해 @RequestMapping 파싱하여 URL을 매핑시키는 기능 구현

- URL마다 하나의 컨트롤러가 할당됩니다. (서블릿과 비슷한 구조)
- Reflection으로 Controller의 RequestMapping Annotation의 URL 정보를 읽어와 Map으로 저장합니다.(URL, Controller)의 key value 구조.

honux77 commented on src/main/java/annotation/RequestMapping.java:
👍

honux77 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,32 @@
+            InvocationTargetException,
+            IllegalAccessException, InstantiationException {
+        Map<String, Method> map = new HashMap<>();
+        map.put("GET", this.getClass().getMethod("doGet", HttpRequest.class));
+        map.put("POST", this.getClass().getMethod("doPost", HttpRequest.class));
```
이 부분 함수형 인터페이스로 담을 수도 있지 않을까요? 어떤게 더 깔끔할지는 잘 모르겠네요 ㅎㅎ.

Gwonwoo-Nam committed:
refactor(step-5):홈 URL 컨트롤러 ArticleController -> HomeController로 변경

Gwonwoo-Nam committed:
feat(step-5):HttpResponse를 Controller에서도 인자를 가져서 접근이 가능하도록 의존성 수정

- Response를 미리 만들어서 Controller에 주입시켜주는 구조로 변경함

Gwonwoo-Nam committed:
refactor(step-5):HttpResponse Builder 패턴으로 생성하도록 리팩터링(가독성 개선)

Gwonwoo-Nam committed:
feat(step-5):UserLoginController 구현, users/login url 매핑

- 세션, 쿠키 기능 구현 예정
- user/login -> users/login으로 변경

honux77 reviewed (changes requested):
Poro 보낸 PR에 URLMapper도 빠져 있고, Mockito의존성도 없는 것 같아요.
다시 한 번 확인 부탁드려요

Gwonwoo-Nam committed:
test(step-5):mock 의존성 추가, 테스트 코드 수정

- Reflection 시 경로에 의존해서 테스트 코드에서는 제대로 동작이 안된다. 어떻게 해결해야할지 고민중입니다.

Gwonwoo-Nam commented:
> Poro 보낸 PR에 URLMapper도 빠져 있고, Mockito의존성도 없는 것 같아요. 다시 한 번 확인 부탁드려요

수정 완료했습니다. 그런데 Reflection을 사용한 테스트에서는 제대로 동작되지 않아 틀리는 테스트가 있어 수정 예정입니다.

Gwonwoo-Nam committed:
refactor(step-5):FrontController의 다룰 handler를 mapping 기능 HandlerMapper 클래스로 분리

- HandlerMapper는 static으로 map 기능 제공

Gwonwoo-Nam committed:
feat(step-5):Custom Annotation으로 컨트롤러에 Method 타입을 명시하는 기능 구현

- Method 타입 표시 시 컨트롤러 생성 시에 reflection으로 annotation이 붙은 메소드를 읽어들여 실행 map에 등록한다.

Gwonwoo-Nam committed:
refactor(step-5):HandlerMapping 초기화 시에만 발생하도록 리팩토링

- 기존은 요청을 처리할 때마다 mapping이 발생해서 비효율적이었다.

Gwonwoo-Nam commented:
추가로 Controller의 메소드 매핑도 @MethodType 어노테이션을 추가해서 자동으로 매핑하도록 구현했습니다. 리플렉션 사용 시  코드를 어떻게 테스트할지가 가장 고민입니다. 그리고 Reflection을 사용하면 컴파일타임에 확인할 수 있는 부분이 최소화되고 런타임에 동작하는 로직이 많아져 디버깅이 까다로워진다는 것도 느꼈습니다.

honux77 reviewed (approved):
수고하셨습니다. 4단계에서 리플렉션 적용하고 리팩토링하느라 고생하셨네요.
흥미로운 코드였어요. 재밌었습니다. 👍
